File: 2-order-process-with-service-workers/csharp/Workers.cs
// Job workers for the Camunda 8 order process: check-inventory, charge-payment,
// and ship-items. Each processes jobs of a specific type and simulates some work.
using Camunda.Orchestration.Sdk;

public static class Workers
{
    public static void Register(CamundaClient client)
    {
        client.CreateJobWorker(
            new JobWorkerConfig
            {
                JobType = "check-inventory",
                JobTimeoutMs = 10_000,
                MaxConcurrentJobs = 5,
                WorkerName = "check-inventory-worker",
            },
            async (job, ct) =>
            {
                Console.WriteLine($"Processing check-inventory job: {job.JobKey}");
                var input = job.GetVariables<InventoryInput>();
                var item = string.IsNullOrEmpty(input?.Item) ? "default-item" : input!.Item;
                Console.WriteLine($"Checking inventory for item: {item}");
                // Simulate checking inventory
                await Task.Delay(2000, ct);
                Console.WriteLine($"check-inventory job completed: {job.JobKey}");
                return new { item = $"{item} allocated" };
            });

        client.CreateJobWorker(
            new JobWorkerConfig
            {
                JobType = "charge-payment",
                JobTimeoutMs = 10_000,
                MaxConcurrentJobs = 5,
                WorkerName = "charge-payment-worker",
            },
            async (job, ct) =>
            {
                Console.WriteLine($"Processing charge-payment job: {job.JobKey}");
                // Simulate some work
                await Task.Delay(2000, ct);
                Console.WriteLine($"charge-payment job completed: {job.JobKey}");
                // Auto-completes with no variables
            });

        client.CreateJobWorker(
            new JobWorkerConfig
            {
                JobType = "ship-items",
                JobTimeoutMs = 10_000,
                MaxConcurrentJobs = 5,
                WorkerName = "ship-items-worker",
            },
            async (job, ct) =>
            {
                Console.WriteLine($"Processing ship-items job: {job.JobKey}");
                // Simulate some work
                await Task.Delay(2000, ct);
                Console.WriteLine($"ship-items job completed: {job.JobKey}");
                // Auto-completes with no variables
            });
    }
}

// DTO describing the input variables for the check-inventory job.
public record InventoryInput(string? Item);

File: 2-order-process-with-service-workers/python/workers.py
"""
This file contains the job workers for the Camunda 8 process.
It defines three job workers: check-inventory, charge-payment, and ship-items.
Each worker processes jobs of a specific type and simulates some work.
"""

import asyncio

from camunda_orchestration_sdk import (
    CamundaAsyncClient,
    ConnectedJobContext,
    WorkerConfig,
)


async def check_inventory(job: ConnectedJobContext) -> dict[str, object]:
    job.log.info(f"Processing check-inventory job: {job.job_key}")
    item = job.variables.to_dict().get("item") or "default-item"
    job.log.info(f"Checking inventory for item: {item}")
    # Simulate checking inventory
    await asyncio.sleep(2)
    job.log.info(f"check-inventory job completed: {job.job_key}")
    return {"item": f"{item} allocated"}


async def charge_payment(job: ConnectedJobContext) -> dict[str, object]:
    job.log.info(f"Processing charge-payment job: {job.job_key}")
    # Simulate some work
    await asyncio.sleep(2)
    job.log.info(f"charge-payment job completed: {job.job_key}")
    return {}


async def ship_items(job: ConnectedJobContext) -> dict[str, object]:
    job.log.info(f"Processing ship-items job: {job.job_key}")
    # Simulate some work
    await asyncio.sleep(2)
    job.log.info(f"ship-items job completed: {job.job_key}")
    return {}


def register_workers(client: CamundaAsyncClient) -> None:
    """Register the three job workers on the given Camunda client."""
    client.create_job_worker(
        config=WorkerConfig(
            job_type="check-inventory",
            job_timeout_milliseconds=10_000,
            max_concurrent_jobs=5,
            worker_name="check-inventory-worker",
        ),
        callback=check_inventory,
    )
    client.create_job_worker(
        config=WorkerConfig(
            job_type="charge-payment",
            job_timeout_milliseconds=10_000,
            max_concurrent_jobs=5,
            worker_name="charge-payment-worker",
        ),
        callback=charge_payment,
    )
    client.create_job_worker(
        config=WorkerConfig(
            job_type="ship-items",
            job_timeout_milliseconds=10_000,
            max_concurrent_jobs=5,
            worker_name="ship-items-worker",
        ),
        callback=ship_items,
    )
